applications/StructuralMechanicsApplication/tests/cpp_tests/test_tangent_tensor_numerical_derivation_utility.cpp

```diff
@@ -16,106 +17,145 @@
-    rStressVector[4] = 0.0;
-    rStressVector[5] = 1.45804e-10;
-
-    rStrainVector = ZeroVector(6);
-    rStrainVector[0] = 0.0;
-    rStrainVector[1] = 0.0;
-    rStrainVector[2] = 8.0e-5;
-    rStrainVector[3] = 0.0;
-    rStrainVector[4] = 0.0;
-    rStrainVector[5] = 1.6941e-21;
-
-    Matrix F = IdentityMatrix(6);
-
-    rValues.SetMaterialProperties(rMaterialProperties);
-    rValues.SetDeformationGradientF(F);
-    rValues.SetStrainVector(rStrainVector);
-    rValues.SetStressVector(rStressVector);
-    rValues.SetOptions(ConstitutiveLawOptions);
-    rValues.SetProcessInfo(CurrentProcessInfo);
-
-    ElasticIsotropic3D ConstitutiveLaw = ElasticIsotropic3D();
-    ElasticIsotropic3D *pConstitutiveLaw = &ConstitutiveLaw;
+    Flags constitutive_law_options;
+    constitutive_law_options.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, false);
+    constitutive_law_options.Set(ConstitutiveLaw::USE_ELEMENT_PROVIDED_STRAIN, false);
+    constitutive_law_options.Set(ConstitutiveLaw::COMPUTE_STRESS, true);
+
+    material_properties.SetValue(YOUNG_MODULUS, 210e9);
+    material_properties.SetValue(POISSON_RATIO, 0.22);
+
+    stress_vector = ZeroVector(6);
+    stress_vector[0] = 5.40984e+06;
+    stress_vector[1] = 5.40984e+06;
+    stress_vector[2] = 1.91803e+07;
+    stress_vector[5] = 1.45804e-10;
+
+    strain_vector = ZeroVector(6);
+    strain_vector[2] = 8.0e-5;
+    strain_vector[5] = 1.6941e-21;
+
+    Matrix deformation_gradient = ZeroMatrix(3, 3);
+    deformation_gradient(0,0) = 1.0;
+    deformation_gradient(1,1) = 1.0;
+    deformation_gradient(2,2) = (1.0 - 8.0e-5);
+    deformation_gradient(2,0) = 0.5 * -1.6941e-21;
+    deformation_gradient(0,2) = 0.5 * -1.6941e-21;
+    const double detF = MathUtils<double>::DetMat(deformation_gradient);
+
+    Matrix tangent_moduli = ZeroMatrix(6, 6);
+
+    cl_configuration_values.SetMaterialProperties(material_properties);
+    cl_configuration_values.SetDeformationGradientF(deformation_gradient);
+    cl_configuration_values.SetDeterminantF(detF);
+    cl_configuration_values.SetStrainVector(strain_vector);
+    cl_configuration_values.SetStressVector(stress_vector);
+    cl_configuration_values.SetOptions(constitutive_law_options);
+    cl_configuration_values.SetProcessInfo(CurrentProcessInfo);
+    cl_configuration_values.SetConstitutiveMatrix(tangent_moduli);
+
+    auto p_constitutive_law = KratosComponents<ConstitutiveLaw>().Get("HyperElastic3DLaw").Clone();
 
     Matrix C = ZeroMatrix(6, 6);
-    rValues.SetConstitutiveMatrix(C);
-    CalculateElasticMatrix(C, rMaterialProperties);
-
-    TangentOperatorCalculatorUtility::CalculateTangentTensor(rValues, pConstitutiveLaw);
-    Matrix &Tangent = rValues.GetConstitutiveMatrix();
-
-    for (int i = 0; i < 6; i++)
-    {
-        for (int j = 0; j < 6; j++)
-        {
-            KRATOS_CHECK_NEAR(C(i, j), Tangent(i, j), 1.0e-3);
+    C = p_constitutive_law->CalculateValue(cl_configuration_values,CONSTITUTIVE_MATRIX_PK2, C);
```

Comment:
there should be a test that computes C even after quite a lot of steps. Just to verify it mantains its "approximate tangent" properties

in order to verify: if 

        double alpha = 1.0;
 
         enew = e + alpha*de //this is the strain
         snew = s(enew) //stress after perturbation
         s = s(e) //stress before perturbation
         des.push_back(alpha*de)
         dscomputed = (snew-s)
         dsestimated = (C*de)
 
        //verify that 
       log( (dsestimated-dscomputed)/dscomputed ) approx 2

this is better (and easier) than evaluating the values of C 1 by 1

note that if the order is not approximately 2, you can scale de, a few times until it happens. If it does not in a few iterations, than you have a a wrong implementation

Reply:
Ok

Reply:
I will add this additionally to the current tests